Code/Tools/FBuild/FBuildCore/Protocol/Client.cpp

```diff
@@ -64,37 +70,32 @@ Client::~Client()
     ShutdownAllConnections();
 
     Thread::CloseHandle( m_Thread );
+
```

Comment:
This code seems a little odd. Can you please provide some details on why this is needed?

Reply:
What range of lines of code? line 73 is a blank line so you probably meant some other lines?

Reply:
Sorry - the block starting underneath this line: Lines 74 to 86 inclusive.

ShutdownAllConnections is a blocking call, so clearing the userdata after that would seem to serve no purpose. With that in mind, I'm unclear what this loop is trying to do.

Reply:
The FDELETE ss line avoids a memory leak I saw.
I am clearing the userdata before that since other FDELETE ss calls you have also cleared userdata and I was following the same pattern.

Thoughts?

Reply:
ShutdownAllConnections() signals all existing connections to terminate and waits for them to do so. During this termination, the socket is closed, the connection information is removed from the TCPConnecitonPool and the OnDisconnect() is called for each connection. 

This sequence is the same as when a connection is lost during the middle of a build. All cleanup should occur in response to the OnDisconnect() event. At any point that a connection is lost, the same cleanup should happen.

If you get to this point and still have connections to free, after all connections should be freed and cleaned up any related allocations, it implies there is a leak somewhere. This loop shouldn't need to be here at all if everything is working correctly.